hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/AbstractTestS3AEncryption.java

```diff
@@ -42,11 +46,24 @@
   protected Configuration createConfiguration() {
     Configuration conf = super.createConfiguration();
     S3ATestUtils.disableFilesystemCaching(conf);
-    conf.set(Constants.SERVER_SIDE_ENCRYPTION_ALGORITHM,
-            getSSEAlgorithm().getMethod());
+    patchConfigurationEncryptionSettings(conf);
     return conf;
   }
 
+  /**
+   * This removes the encryption settings from the
```

Comment:
This does not "removes the encryption settings from the configuration". Could you explain why that is not required in branch-3.2 and/or update the java doc here?

Reply:
Yes comment is misleading. I had to fix the merge conflict and as removeBaseAndBucketOverrides() was not present earlier I had to delete that line thinking it is not required. 

Although I can see the same method is present in S3ATestUtils but with different parameters and it is not used in this test. 
@steveloughran  Do you also think tests might break here if bucket level conf are set?

Reply:
it may well

Reply:
I tried setting the bucket level conf  for fs.s3a.bucket.mthakur-data.server-side-encryption.key and some tests for eg. ITestS3AEncryptionSSES3 started failing. I even tried adding the removeBaseAndBucketOverrides call but it doesn't help. 

I tried setting then same bucket level conf  in apache/trunk as well and the tests are succeeding there. 

The reason for test failure is 
`java.io.IOException: AES256 is enabled but an encryption key was set in fs.s3a.server-side-encryption.key (key of length 76 ending with 8)
	at org.apache.hadoop.fs.s3a.S3AUtils.getEncryptionAlgorithm(S3AUtils.java:1440)
	at org.apache.hadoop.fs.s3a.S3AFileSystem.initialize(S3AFileSystem.java:316)
	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:3298)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:476)` 

removeBaseAndBucketOverrides() call essentially removes this parameter not sure how it is getting added in FileSystem initialisation flow again. It is too hard to compare and debug the difference between branch-3.2 and trunk as there are many changes. Any pointers in this regards would be greatly appreciated. 

Also could someone please trigger the test based on their configs and share the results. Thanks.

Reply:
I can reproduce the test failure. I ran the test with `fs.s3a.bucket.BUCKETNAME.server-side-encryption.key` set in `auth-keys.xml`.

I set the S3 bucket default encryption policy as SSE-KMS, and that is not changing the `ITestS3AEncryptionSSES3 ` test failure (as expected). Well, test `ITestS3AEncryptionWithDefaultS3Settings` could pass then (again, as expected).

Reply:
OK, I did some investigation, and the observation is as following:

1. The failing tests are just about tests. The fix in copy options should work as expected. Specially, some encryption tests fail only when bucket level encryption is set.
1. The inconsistent encryption algorithm and key situation in conf is *partially* because of the `addConfResource(CONTRACT_XML);` function after creating and patching the conf object, see [here](https://github.com/apache/hadoop/blob/9cad3e235026dbe4658705ca85d263d0edf14521/hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/contract/s3a/S3AContract.java#L36).
1. The call to `addConfResource(CONTRACT_XML);` should be considered a tricky unrelated bug, not brought by this patch. Most likely it is fixed by HADOOP-16626 in {{trunk}}. Guess that's why we don't see test failures there. I'm wondering how {{trunk}} is not failing this...? Was HADOOP-16626 enough? 
1. Specially, there are some discussions about the Configuration loading resources "after" the conf object has been patched when setting up the S3 tests. In our test case, if you have the change to call `removeBucketOverrides()`, and you have the change in `S3AContract` not calling `addConfResource(CONTRACT_XML);`, then you can test with this: you add multiple `testEncryption()` and multiple `testEncryptionOverRename()` test cases in `AbstractTestS3AEncryption`. Guess what, only one test fail and all others are failing...How interesting? This seems very related to static resource loading when creating S3AFileSystem...

I'll stop here since I for now have no more context. I will request @steveloughran to provide more help. Shall we backport HADOOP-16626 to {{branch-3.2}} and see?